orttraining/orttraining/python/training/onnxblock/building_blocks.py

```diff
@@ -0,0 +1,221 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# _building_blocks.py
+
+from abc import ABC, abstractmethod
+import onnx
+
+import onnxruntime.training.onnxblock.model_accessor as accessor
+import onnxruntime.training.onnxblock._graph_utils as graph_utils
+
+
+class Block(ABC):
+    """Base class for all building blocks that can be stacked on top of each other."""
+
+    def __init__(self):
+        ...
+
+    @abstractmethod
+    def build(self, *args, **kwargs):
+        """Customize and build this block.
+
+        This method is to be overriden by the user's implementation.
+        """
+        ...
+
+    def __call__(self, *args, **kwargs):
+        """Calls the user's build method and runs validation on top."""
+
+        # build the user model
+        output = self.build(*args, **kwargs)
+
+        # validate and check the model
+        onnx.checker.check_model(accessor.global_accessor.model, True)
+
+        return output
+
+
+class _BinaryOp(Block):
+    def __init__(self):
+        super(_BinaryOp, self).__init__()
+
+    def _build(self, input_name1, input_name2, op_name):
+        # get the model to manipulate
+        onnx_model = accessor.global_accessor.model
+
+        # create the graph node for sub
+        node_input_names = [input_name1, input_name2]
+        node_output_name = graph_utils.generate_random_graph_name(
+            f"{op_name.lower()}_output"
+        )
+        node_output_names = [node_output_name]
+        node = onnx.helper.make_node(
+            op_name,
+            node_input_names,
+            node_output_names,
+            name=graph_utils.generate_random_graph_name(op_name),
+        )
+        onnx_model.graph.node.append(node)
+
+        return node_output_name
+
+
+class Add(_BinaryOp):
```

Comment:
Do we need a separate class for each binary op? Can we simply call the build mtd of BinaryOp with the op_name?

Reply:
I can potentially make this simpler, but I don't want users to instantiate an instance of `BinaryOp` using `add = BinadryOp("Add")`. This kind of instantiation in my opinion is not very programmatic and can lead to more string mistyping errors.
I think it is better to have a class for each class.